src/main/scala/mesosphere/marathon/core/task/termination/impl/KillStreamWatcher.scala

```diff
@@ -1,101 +1,112 @@
+    * or if the instance itself is in a state we consider terminal.
     */
-  private def killedInstanceIds(eventStream: akka.event.EventStream, instancesToCheck: Iterable[Instance]): Source[Instance.Id, Cancellable] = {
-    // MaxValue causes a dynamically sized buffer to be used.
-    val bufferSize = Int.MaxValue
-    val overflowStrategy = OverflowStrategy.fail
+  private[impl] val whenTerminalOrReplaced: TerminalPredicate = (instance: Instance, taskIds: Set[Task.Id]) =>
+    considerTerminal(instance.state.condition) || instance.tasksMap.values.forall(t => !taskIds(t.taskId))
 
-    val alreadyConsideredTerminal = instancesToCheck
-      .filter(instance => considerTerminal(instance.state.condition))
-      .map(_.instanceId)
-      .toList
+  /**
+    * Predicate used to determine whether or not an instance has been decommissioned.
+    *
+    * When an instance has been marked as goal decommissioned, it's safe to assume that it is never coming back, since
+    * it's illegal to change an instance goal from Decommissioned back to running. As such, we assume an instance is
+    * decommissioned as soon as it is both marked as goal decommissioned, and the instance is considered terminal.
+    *
+    * Were we to only wait for the instance to become terminal, and not for the goal to also be decommissioned, then we
+    * could conclude too early that an instance is decommissioned, as it is possible that some random restart could
+    * precede the receipt of the update goal command.
+    */
+  private[impl] val whenTerminalAndDecommissioned: TerminalPredicate = (instance: Instance, _: Set[Task.Id]) =>
+    instance.state.goal == Goal.Decommissioned && considerTerminal(instance.state.condition)
 
-    val killedViaInstanceChanged =
-      EnrichedSource.eventBusSource(classOf[InstanceChanged], eventStream, bufferSize, overflowStrategy).collect {
-        case event if considerTerminal(event.condition) => event.id
-      }
+  /**
+    * Given instance updates, and a set of instances, emit a set of instanceIds that are not yet expunged or considered
+    * terminal. Excludes instances that are expunged or considered terminal in the initial snapshot.
+    *
+    * @param instanceUpdates InstanceTracker instance state subscription stream
+    * @param instances The instances in question to watch
+    * @param predicate The mechanism to judge instance terminality; are we watching for tasks to be killed, or instance to be decommissioned.
+    * @return
+    */
+  private[impl] def emitPendingTerminal(instanceUpdates: InstanceTracker.InstanceUpdates, instances: Iterable[Instance], predicate: TerminalPredicate): Source[Set[Instance.Id], NotUsed] = {
 
-    val killedViaUnknownInstanceTerminated =
-      EnrichedSource.eventBusSource(
-        classOf[UnknownInstanceTerminated], eventStream, bufferSize, overflowStrategy).map(_.id)
+    val instancesMap: Map[Instance.Id, Set[Task.Id]] =
```

Comment:
Nit: I was confused by the naming. What do you think about `instancesTasks`? Or maybe even more verbose `instancesToInitialTasks`?

Reply:
Yeah, good point, I'll make the change